######################################
#
# BUSCO summary figure
# @version 4.0.0
# @since BUSCO 2.0.0
# 
# Copyright (c) 2016-2022, Evgeny Zdobnov (ez@ezlab.org)
# Licensed under the MIT license. See LICENSE.md file.
#
######################################

# Load the required libraries
library(ggplot2)
library("grid")


# !!! CONFIGURE YOUR PLOT HERE !!! 
# Output
my_output <- paste("~/@Uni/@Flinders University PhD/RWorkingDir/BUSCO/BUSCO_summaries","busco_figure.png",sep="/") 
my_width <- 20
my_height <- 15
my_unit <- "cm"

# Colors
my_colors <- c("#56B4E9", "#3492C7", "#F0E442", "#F04442")
# Bar height ratio
my_bar_height <- 0.75

# Legend
my_title <- "BUSCO Assessment Results"

# Font
my_family <- "sans"
my_size_ratio <- 1

# !!! SEE YOUR DATA HERE !!! 
# "(SR) G1_SepF" "(SR) G2_SepF" "(SR) G3_MarF" "(SR) G4_MarM" "(SR) G5_SepM" "(SR) G6_SepM" "(SR) G7_MarF" "(SR) G8_MarF" "1 (LR) High Quality Unique Isoforms" "2 (LR) Unique Isoform Coding Sequence" "3 (LR) Clustered Transcripts: Putative Genes" "4 (SR) Combined TRINITY Assembly"
# Your data as generated by python, remove or add more
my_species <- c('(SR) G1_SepF', '(SR) G1_SepF', '(SR) G1_SepF', '(SR) G1_SepF', '(SR) G2_SepF', '(SR) G2_SepF', '(SR) G2_SepF', '(SR) G2_SepF', '(SR) G3_MarF', '(SR) G3_MarF', '(SR) G3_MarF', '(SR) G3_MarF', '(SR) G4_MarM', '(SR) G4_MarM', '(SR) G4_MarM', '(SR) G4_MarM', '(SR) G5_SepM', '(SR) G5_SepM', '(SR) G5_SepM', '(SR) G5_SepM', '(SR) G6_SepM', '(SR) G6_SepM', '(SR) G6_SepM', '(SR) G6_SepM', '(SR) G7_MarF', '(SR) G7_MarF', '(SR) G7_MarF', '(SR) G7_MarF', '(SR) G8_MarF', '(SR) G8_MarF', '(SR) G8_MarF', '(SR) G8_MarF', '3 (LR) Clustered Transcripts: Putative Genes', '3 (LR) Clustered Transcripts: Putative Genes', '3 (LR) Clustered Transcripts: Putative Genes', '3 (LR) Clustered Transcripts: Putative Genes', '2 (LR) Unique Isoform Coding Sequence', '2 (LR) Unique Isoform Coding Sequence', '2 (LR) Unique Isoform Coding Sequence', '2 (LR) Unique Isoform Coding Sequence', '1 (LR) High Quality Unique Isoforms', '1 (LR) High Quality Unique Isoforms', '1 (LR) High Quality Unique Isoforms', '1 (LR) High Quality Unique Isoforms', '0 (SR) Combined TRINITY Assembly', '0 (SR) Combined TRINITY Assembly', '0 (SR) Combined TRINITY Assembly', '0 (SR) Combined TRINITY Assembly')
my_species <- factor(my_species)
my_species <- factor(my_species,levels(my_species)[c(length(levels(my_species)):1)]) # reorder your species here just by changing the values in the vector :
my_percentage <- c(48.4, 25.3, 11.1, 15.2, 49.8, 20.9, 11.6, 17.7, 43.9, 14.8, 13.4, 27.9, 43.1, 19.5, 13.3, 24.1, 45.2, 15.1, 15.1, 24.6, 46.3, 44.5, 3.3, 5.9, 34.4, 50.9, 5.5, 9.2, 45.5, 20.9, 11.7, 21.9, 44.9, 7.3, 2.9, 44.9, 31.6, 15.5, 6.1, 46.8, 35.0, 17.8, 2.9, 44.3, 30.3, 64.6, 1.8, 3.3)
my_values <- c(1625, 847, 371, 511, 1670, 701, 390, 593, 1474, 498, 448, 934, 1445, 654, 445, 810, 1516, 508, 507, 823, 1552, 1493, 110, 199, 1154, 1707, 184, 309, 1527, 700, 391, 736, 1507, 244, 96, 1507, 1059, 519, 203, 1573, 1174, 597, 97, 1486, 1017, 2168, 62, 107)

######################################
######################################
######################################
# Code to produce the graph
labsize = 1
if (length(levels(my_species)) > 10){
 labsize = 0.66
}
print("Plotting the figure ...")
category <- c(rep(c("S","D","F","M"),c(1)))
category <-factor(category)
category = factor(category,levels(category)[c(4,1,2,3)])
df = data.frame(my_species,my_percentage,my_values,category)

figure <- ggplot() + 
  
  geom_bar(aes(y = my_percentage, x = my_species, fill = category), position = position_stack(reverse = TRUE), data = df, stat="identity", width=my_bar_height) + 
  coord_flip() + 
  theme_gray(base_size = 8) + 
  scale_y_continuous(labels = c("0","20","40","60","80","100"), breaks = c(0,20,40,60,80,100)) + 
  scale_fill_manual(values = my_colors,labels =c(" Complete (C) and single-copy (S)  ",
                                                 " Complete (C) and duplicated (D)",
                                                 " Fragmented (F)  ",
                                                 " Missing (M)")) +   
  ggtitle(my_title) + 
  xlab("") + 
  ylab("\n%BUSCOs") + 

  theme(plot.title = element_text(family=my_family, hjust=0.5, colour = "black", size = rel(2.2)*my_size_ratio, face = "bold")) + 
  theme(legend.position="top",legend.title = element_blank()) + 
  theme(legend.text = element_text(family=my_family, size = rel(1.2)*my_size_ratio)) + 
  theme(panel.background = element_rect(color="#FFFFFF", fill="white")) + 
  theme(panel.grid.minor = element_blank()) + 
  theme(panel.grid.major = element_blank()) +
  theme(axis.text.y = element_text(family=my_family, colour = "black", size = rel(1.66)*my_size_ratio)) + 
  theme(axis.text.x = element_text(family=my_family, colour = "black", size = rel(1.66)*my_size_ratio)) + 
  theme(axis.line = element_line(size=1*my_size_ratio, colour = "black")) + 
  theme(axis.ticks.length = unit(.85, "cm")) + 
  theme(axis.ticks.y = element_line(colour="white", size = 0)) + 
  theme(axis.ticks.x = element_line(colour="#222222")) + 
  theme(axis.ticks.length = unit(0.4, "cm")) + 
  theme(axis.title.x = element_text(family=my_family, size=rel(1.2)*my_size_ratio)) + 
  
  guides(fill = guide_legend(override.aes = list(colour = NULL))) +
  guides(fill=guide_legend(nrow=2,byrow=TRUE))
  
  for(i in rev(c(1:length(levels(my_species))))){
    detailed_values <- my_values[my_species==my_species[my_species==levels(my_species)[i]]]
    total_buscos <- sum(detailed_values)
    figure <- figure + 
    annotate("text", label=paste("C:", detailed_values[1] + detailed_values[2], " [S:", detailed_values[1], ", D:", detailed_values[2], "], F:", detailed_values[3], ", M:", detailed_values[4], ", n:", total_buscos, sep=""), 
             y=3, x = i, size = labsize*4*my_size_ratio, colour = "black", hjust=0, family=my_family)
  }
  
ggsave(figure, file=my_output, width = my_width, height = my_height, unit = my_unit)
print("Done")
